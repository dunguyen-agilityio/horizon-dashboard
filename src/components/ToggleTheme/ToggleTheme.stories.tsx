import type { Meta, StoryObj } from '@storybook/react';
import ToggleTheme from '.';

const meta = {
  title: 'Components/ToggleTheme',
  component: ToggleTheme,
  tags: ['autodocs'],
} satisfies Meta<typeof ToggleTheme>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Primary: Story = {
  args: {},
};

export const Secondary: Story = {
  args: {
    variant: 'secondary',
  },
};
