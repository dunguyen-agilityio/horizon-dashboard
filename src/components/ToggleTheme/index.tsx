'use client';

import { useTheme } from 'next-themes';
import { useEffect, useState } from 'react';

// Components
import Moon from '@/icons/Moon';
import { cn } from '@nextui-org/theme';
import Button from '../Button';

interface ToggleThemeProps {
  className?: string;
  variant?: 'primary' | 'secondary';
}

const ToggleTheme = ({
  className,
  variant = 'primary',
  ...rest
}: ToggleThemeProps) => {
  const { theme, setTheme } = useTheme();
  const [isMounted, setIsMounted] = useState(false);

  useEffect(() => {
    setIsMounted(true);
  }, []);

  if (!isMounted) return null;

  const isDark = theme === 'dark';

  const handleChangeTheme = () => {
    setTheme(isDark ? 'light' : 'dark');
  };

  const isPrimary = variant === 'primary';

  return (
    <Button
      size="fit"
      onClick={handleChangeTheme}
      className={cn(
        'rounded-full flex justify-center items-center',
        isPrimary
          ? 'p-2 bg-transparent'
          : 'h-[60px] min-w-[60px] bg-gradient-primary',
        className,
      )}
      variant={isPrimary ? undefined : 'light'}
      {...rest}
    >
      <Moon isDark={isPrimary ? isDark : true} />
    </Button>
  );
};

export default ToggleTheme;
