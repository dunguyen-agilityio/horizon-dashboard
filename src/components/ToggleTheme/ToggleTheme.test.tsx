import { useTheme } from 'next-themes';
import { act, fireEvent, render, waitFor } from '@testing-library/react';

import ToggleTheme from '.';

jest.mock('next-themes', () => ({
  useTheme: jest.fn(),
}));

describe('ToggleTheme tests', () => {
  let mockSetTheme: jest.Mock;

  beforeEach(() => {
    mockSetTheme = jest.fn();
    (useTheme as jest.Mock).mockImplementation(() => ({
      theme: 'light',
      setTheme: mockSetTheme,
    }));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('Should match snapshot', () => {
    const { container } = render(<ToggleTheme />);
    expect(container).toMatchSnapshot();
  });

  it('Update theme will called correctly', async () => {
    const { getByTestId } = render(
      <ToggleTheme variant="secondary" data-testid="toggle-theme" />,
    );
    act(() => fireEvent.click(getByTestId('toggle-theme')));
    await waitFor(() =>
      expect(mockSetTheme).toHaveBeenNthCalledWith(1, 'dark'),
    );
  });

  it('Update theme will called correctly', async () => {
    (useTheme as jest.Mock).mockImplementation(() => ({
      theme: 'dark',
      setTheme: mockSetTheme,
    }));

    const { getByTestId } = render(<ToggleTheme data-testid="toggle-theme" />);
    act(() => fireEvent.click(getByTestId('toggle-theme')));
    await waitFor(() =>
      expect(mockSetTheme).toHaveBeenNthCalledWith(1, 'light'),
    );
  });
});
