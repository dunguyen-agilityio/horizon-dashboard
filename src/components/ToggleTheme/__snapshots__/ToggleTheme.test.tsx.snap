// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleTheme tests Should match snapshot 1`] = `
<div>
  <button
    class="z-0 group relative box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-small gap-2 [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none text-default-foreground data-[hover=true]:opacity-hover min-w-fit h-fit rounded-full flex justify-center items-center p-2 bg-transparent"
    type="button"
  >
    <svg
      class="[&_path]:fill-secondary dark:[&_path]:fill-white [&_rect]:fill-secondary dark:[&_rect]:fill-white"
      data-testid="moon-icon"
      fill="none"
      height="18"
      viewBox="0 0 18 18"
      width="18"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        clip-path="url(#clip0_101_12742)"
      >
        <path
          d="M9.95703 18C12.733 18 15.2684 16.737 16.9481 14.6675C17.1966 14.3613 16.9256 13.9141 16.5416 13.9872C12.1751 14.8188 8.16522 11.4709 8.16522 7.06303C8.16522 4.52398 9.52444 2.18914 11.7335 0.931992C12.074 0.738211 11.9884 0.221941 11.6015 0.150469C11.059 0.0504468 10.5086 8.21369e-05 9.95703 0C4.98914 0 0.957031 4.02578 0.957031 9C0.957031 13.9679 4.98281 18 9.95703 18Z"
        />
      </g>
      <defs>
        <clippath
          id="clip0_101_12742"
        >
          <rect
            height="18"
            width="18"
          />
        </clippath>
      </defs>
    </svg>
  </button>
</div>
`;
